clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
@MayureshGharat in general we may have some requests that timed out. Thus we have to remove the specific object from the waiters queue, right? To do that, we probably need to replace `Deque` with something like `ConcurrentLinkedDeque` for waiters, or use lock to protect waiters.